src/ServiceControl.Audit.Persistence.PostgreSQL/UnitOfWork/PostgreSQLAuditIngestionUnitOfWork.cs

```diff
@@ -0,0 +1,143 @@
+        cmd.Parameters.AddWithValue("message_type", GetMetadataOrDbNull<string>("MessageType"));
+        cmd.Parameters.AddWithValue("is_system_message", GetMetadataOrDbNull<bool>("IsSystemMessage"));
+        cmd.Parameters.AddWithValue("time_sent", GetMetadataOrDbNull<DateTime>("TimeSent"));
+        cmd.Parameters.AddWithValue("receiving_endpoint_name", GetMetadata<EndpointDetails>("ReceivingEndpoint")?.Name ?? (object)DBNull.Value);
+        cmd.Parameters.AddWithValue("critical_time", GetMetadataOrDbNull<TimeSpan>("CriticalTime"));
+        cmd.Parameters.AddWithValue("processing_time", GetMetadataOrDbNull<TimeSpan>("ProcessingTime"));
+        cmd.Parameters.AddWithValue("delivery_time", GetMetadataOrDbNull<TimeSpan>("DeliveryTime"));
+        cmd.Parameters.AddWithValue("conversation_id", GetMetadataOrDbNull<string>("ConversationId"));
+        cmd.Parameters.AddWithValue("status", (int)(GetMetadata<bool>("IsRetried") ? MessageStatus.ResolvedSuccessfully : MessageStatus.Successful));
+
+        batch.BatchCommands.Add(cmd);
+        return Task.CompletedTask;
+    }
+
+    public Task RecordSagaSnapshot(SagaSnapshot sagaSnapshot, CancellationToken cancellationToken)
+    {
+        var newChange = new
+        {
+            sagaSnapshot.StartTime,
+            sagaSnapshot.FinishTime,
+            sagaSnapshot.Status,
+            sagaSnapshot.StateAfterChange,
+            sagaSnapshot.InitiatingMessage,
+            sagaSnapshot.OutgoingMessages,
+            sagaSnapshot.Endpoint
+        };
+
+        // Insert or update saga_snapshots table - add new change to the changes array
+        var cmd = batch.CreateBatchCommand();
+        cmd.CommandText = @"
+                INSERT INTO saga_snapshots (id, saga_id, saga_type, changes)
+                VALUES (@saga_id, @saga_id, @saga_type, @new_change)
+                ON CONFLICT (id) DO UPDATE SET
+                    changes = COALESCE(saga_snapshots.changes, '[]'::jsonb) || @new_change::jsonb;";
+
+        cmd.Parameters.AddWithValue("saga_id", sagaSnapshot.SagaId);
+        cmd.Parameters.AddWithValue("saga_type", sagaSnapshot.SagaType);
+        cmd.Parameters.AddWithValue("new_change", NpgsqlTypes.NpgsqlDbType.Jsonb, new[] { newChange });
+
+        batch.BatchCommands.Add(cmd);
+
+        return Task.CompletedTask;
+    }
+
+    public Task RecordKnownEndpoint(KnownEndpoint knownEndpoint, CancellationToken cancellationToken)
+    {
+        // Insert KnownEndpoint into known_endpoints table
+        var cmd = batch.CreateBatchCommand();
+        cmd.CommandText = @"
+                INSERT INTO known_endpoints (
+                    id, name, host_id, host, last_seen
+                ) VALUES (
+                    @id, @name, @host_id, @host, @last_seen
+                )
+                ON CONFLICT (id) DO UPDATE SET
+                    last_seen = GREATEST(known_endpoints.last_seen, EXCLUDED.last_seen);";
```

Comment:
After running a test scaling out the audit instance, this query caused too many deadlocks.
This would have to be redesigned to prevent that.